### ISSUE TITLE
Add support for CallableStatement

### ISSUE THREAD

**Issue by hsn10** [scalikejdbc/scalikejdbc#438]
I would like to have support for calling stored procedures with in/out parameters. Example of Java JDBC code:

            con = getDataSource().getConnection();
            CallableStatement callable = con.prepareCall("CALL SYSPROC.ADMIN_DROP_SCHEMA(?, ?, ?, ?)");
            callable.setString(1, dbSettings.getDatabaseSchema().toUpperCase());
            callable.setNull(2, Types.VARCHAR);
            callable.registerOutParameter(3, Types.VARCHAR);
            callable.registerOutParameter(4, Types.VARCHAR);
            callable.setString(3, "SYSTOOLS");
            callable.setString(4, "DROPSCHEMA");
            callable.executeUpdate();

 i propose to add some method like call(template: String, params: CallableArg*): Any to DBSession

And callableArg to be structure holding val type = java.sql.Types.* and var value and val parameter type - in/out/inout

**Comment by seratch** [comment]
Although I've considered supporting CallableStatement before, I've not figured out what the best API design is. 

Related issues:

- https://github.com/scalikejdbc/scalikejdbc/issues/59 CallableStatement support
- https://github.com/scalikejdbc/scalikejdbc/issues/380 callable statements API: Need to be able to call stored functions
- https://github.com/scalikejdbc/scalikejdbc/issues/336 How to call procedures and get return value? 

At this time, doing like this is the most flexible and handy way to use CallableStatement.

https://github.com/scalikejdbc/scalikejdbc/issues/63#issuecomment-12502620

Your suggestion basically looks reasonable. To be honest, I myself don't know many use cases of CallableStatement. I'd love to see realistic examples more.

**Comment by hsn10** [comment]
Realistic use case of CallableStatement is to call stored procedure, Any more advanced application has stored procedures for security or performance reasons.

**Comment by seratch** [comment]
Yeah, that's right. I meant having realistic test cases will help us to make the API design mature.

**Comment by VEINHORN** [comment]
@seratch Hi. Are there any plans to add support for CallableStatement in near future? I'm forced to mix Scala/Java for this purposes. :/

**Comment by seratch** [comment]
@VEINHORN To be honest, I'm not in a hurry because I myself have less opportunities to use CallableStatement at least now. So I have no confidence about its appropriate API design. Do you have any suggestions for appropriate API design? If you have nice idea, please let me know.

**Comment by andrea-mariotti** [comment]
One more vote for CallableStatement support, we use stored procedure in Postgres and it would help a lot. Please and thank you.

**Comment by webgress** [comment]
CallableStatement  is the deal breaker.
If I knew it was missing would not have spend the time using the lib.
It was nice but now has to be removed.
As was mentioned earlier - some DBAs will allow only stored procedure access.

**Comment by seratch** [comment]
This is still an open issue. I myself do not have the bandwidth to work on this feature in the short term but contributions are welcome. If you are interested in working on this, please share the design (classes, methods, how to use) before sending a pull request with complete code.
